core/client/fs/src/main/java/alluxio/client/block/stream/SharedGrpcDataReader.java

```diff
@@ -0,0 +1,187 @@
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * A shared gRPC data reader that cache blocks data for multi-thread accessing.
+ *
+ * It follows GrpcDataReader protocol and takes strong assumption:
+ * Parallel read to the same file happens on the same time, so that read request is
+ * serialized by kernel
+ */
+@NotThreadSafe
+public class SharedGrpcDataReader implements DataReader {
+  private static final Logger LOG = LoggerFactory.getLogger(SharedGrpcDataReader.class);
+  private static final ReentrantReadWriteLock[] BLOCK_LOCKS;
+  /** A map from block id to the block's cached data reader. */
+  @GuardedBy("BLOCK_LOCKS")
+  private static final ConcurrentHashMap<Long, BufferCachingGrpcDataReader> BLOCK_READERS
+      = new ConcurrentHashMap<>();
+
+  static {
+    AlluxioConfiguration conf = new InstancedConfiguration(ConfigurationUtils.defaults());
+    int lockNum = conf.getInt(PropertyKey.FUSE_SHARED_CACHING_READER_LOCKS);
+    BLOCK_LOCKS = new ReentrantReadWriteLock[lockNum];
+    for (int i = 0; i < lockNum; i++) {
+      BLOCK_LOCKS[i] = new ReentrantReadWriteLock();
+    }
+  }
+
+  private final long mBlockId;
+  private final BufferCachingGrpcDataReader mCachedDataReader;
+  private final long mChunkSize;
+
+  /** The next pos to read. */
+  private long mPosToRead;
+
+  /**
+   * Creates an instance of {@link SharedGrpcDataReader}.
+   *
+   * @param readRequest the read request
+   * @param reader the cached Grpc data reader for the given block
+   */
+  private SharedGrpcDataReader(ReadRequest readRequest, BufferCachingGrpcDataReader reader) {
+    mChunkSize = readRequest.getChunkSize();
+    mPosToRead = readRequest.getOffset();
+    mBlockId = readRequest.getBlockId();
+    mCachedDataReader = reader;
+  }
+
+  @Override
+  public long pos() {
+    return mPosToRead;
+  }
+
+  /**
+   * Seeks to a specific position.
+   *
+   * @param pos the position to seek to
+   */
+  public void seek(long pos) {
+    mPosToRead = pos;
+  }
+
+  @Override
+  public DataBuffer readChunk() throws IOException {
+    int index = (int) (mPosToRead / mChunkSize);
+    DataBuffer chunk = mCachedDataReader.readChunk(index);
+    if (chunk == null) {
+      return null;
+    }
+    ByteBuffer bb = chunk.getReadOnlyByteBuffer();
+    // Force to align to chunk size
+    bb.position((int) (mPosToRead % mChunkSize));
+    mPosToRead += mChunkSize - mPosToRead % mChunkSize;
+
+    return new NioDataBuffer(bb, bb.remaining());
+  }
+
+  @Override
+  public DataBuffer readChunkIfReady() throws IOException {
+    // I'm naive, I'm reading chunks anyway
+    return readChunk();
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (mCachedDataReader.deRef() == 0) {
+      try (LockResource lockResource = new LockResource(
+          BLOCK_LOCKS[(int) (mBlockId % BLOCK_LOCKS.length)].writeLock())) {
```

Comment:
let's make a static util method to return the corresponding lock in the lock array based on block Id.
e.g.,  
```
private static ReentrantReadWriteLock getBlockLock(long blockId) {
  return BLOCK_LOCKS[(int) (blockId % BLOCK_LOCKS.length)]
}
```
Then in both `close` and `create`, let's all use the util method rather than hardcode the same logic twice. 
In case we change the logic in the future, we only need to go one place

Reply:
Changed, thanks